tests/unit/test_add_credential.py

```diff
@@ -0,0 +1,65 @@
+import pathlib
+
+import pytest
+
+import jubilant
+
+from . import mocks
+
+
+def test_with_path_str(run: mocks.Run):
+    run.handle(['juju', 'add-credential', 'aws', '--client', '--file', '/path/to/creds.yaml'])
+    juju = jubilant.Juju()
+
+    juju.add_credential('aws', '/path/to/creds.yaml', client=True)
+
+
+def test_with_path_pathlib(run: mocks.Run):
+    run.handle(['juju', 'add-credential', 'aws', '--client', '--file', '/path/to/creds.yaml'])
+    juju = jubilant.Juju()
+
+    juju.add_credential('aws', pathlib.Path('/path/to/creds.yaml'), client=True)
+
+
+def test_with_yaml_dict(run: mocks.Run, mock_file: mocks.NamedTemporaryFile):
+    run.handle(['juju', 'add-credential', 'aws', '--controller', 'cc', '--file', mock_file.name])
+    juju = jubilant.Juju()
+
+    credential = {'credentials': {'aws': {'mycred': {'auth-type': 'access-key'}}}}
+    juju.add_credential('aws', credential, controller='cc')
+
+    assert len(mock_file.writes) == 1
+    assert 'auth-type: access-key' in mock_file.writes[0]
+
+
+def test_all_args(run: mocks.Run):
+    run.handle(
+        [
+            'juju',
+            'add-credential',
+            'aws',
+            '--client',
+            '--controller',
+            'mycontroller',
+            '--region',
+            'us-east-1',
+            '--file',
+            '/path/to/creds.yaml',
+        ]
+    )
+    juju = jubilant.Juju()
+
+    juju.add_credential(
+        'aws',
+        credential='/path/to/creds.yaml',
+        client=True,
+        controller='mycontroller',
+        region='us-east-1',
+    )
+
+
+def test_neither_client_nor_controller():
+    juju = jubilant.Juju()
+
+    with pytest.raises(TypeError):
+        juju.add_credential('aws', '/path/to/creds.yaml')  # type: ignore
```

Comment:
IMO the fact that we got the overloads wrong the first time indicates that it would be worth having exhaustive coverage of the relevant argument combinations.

My personal preference is to have this in a dedicated file for code that is only exercised by the type checker, as that lets you enumerate the cases more tersely, but we could still do it in unit tests, though it's a bit noisy as you mentioned.

That said, I guess we have verified that it's correct for now through manual testing, so it's not strictly necessary.
```suggestion
    with pytest.raises(TypeError):
        juju.add_credential('aws', 'creds.yaml')  # type: ignore
    with pytest.raises(TypeError):
        juju.add_credential('aws', 'creds.yaml', client=False)  # type: ignore
    with pytest.raises(TypeError):
        juju.add_credential('aws', 'creds.yaml', controller=None)  # type: ignore
    with pytest.raises(TypeError):
        juju.add_credential('aws', 'creds.yaml', client=False, controller=None)  # type: ignore
```